src/bin/main.rs

```diff
@@ -228,6 +199,152 @@ fn deadbeef (args: Args) -> Result<(), Error> {
+                // null
+                0x00, 0x00,
+
+                // null
+                0x00,
+            ])?;
+        let mut debug_info =
+            vec![
+                // 0x00: Length = 0x34 - 4
+                0x30, 0x00, 0x00, 0x00,
+                // 0x04: Version
+                0x04, 0x00,
+                // 0x06: Abbrev offset (needs reloc)
+                0x00, 0x00, 0x00, 0x00,
+                // 0x0a: Address size
+                0x08,
+
+                // 0x0b: Abbrev 1 = DW_TAG_compile_unit
+                0x01,
+                // 0x0c: DW_AT_producer = 0x00 (needs reloc)
+                0x00, 0x00, 0x00, 0x00,
+                // 0x10: DW_AT_language = DW_LANG_C
+                0x02,
+                // 0x11: DW_AT_name = 0x19 (needs reloc)
+                0x00, 0x00, 0x00, 0x00,
+                // 0x15: DW_AT_comp_dir = 0x07 (needs reloc)
+                0x00, 0x00, 0x00, 0x00,
+
+                // 0x19: Abbrev 2 = DW_TAG_variable
+                0x02,
+                // 0x1a: DW_AT_name = 0x24 (needs reloc)
+                0x00, 0x00, 0x00, 0x00,
+                // 0x1e: DW_AT_type = offset of int base_type
+                0x2c, 0x00, 0x00, 0x00,
+                // 0x22: DW_FORM_flag_present = no data needed
+                // 0x22: DW_AT_location = len 9, DW_OP_addr DEADBEEF (needs reloc)
+                0x09, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+
+                // 0x2c: Abbrev 3 = DW_TAG_base_type
+                0x03,
+                // 0x2d: DW_AT_name = "int"
+                b'i', b'n', b't', 0x00,
+                // 0x31: DW_AT_byte_size = 4
+                0x04,
+                // 0x32: DW_AT_encoding = DW_ATE_signed
+                0x05,
+
+                // 0x33: End of children
+                0x00,
+            ];
+
+        if args.mach {
+            // No relocation needed for Mach.
+            debug_info[0x11] = 0x19;
+            debug_info[0x15] = 0x7;
+            debug_info[0x1a] = 0x24;
+        } else {
+            // abbrev offset
+            obj.link_with(
+                Link { from: ".debug_info", to: ".debug_abbrev", at: 0x06},
+                Reloc::Debug { size: 4, addend: 0x0},
+            )?;
+            // producer
+            obj.link_with(
+                Link { from: ".debug_info", to: ".debug_str", at: 0x0c},
+                Reloc::Debug { size: 4, addend: 0x0},
+            )?;
+            // CU name
+            obj.link_with(
+                Link { from: ".debug_info", to: ".debug_str", at: 0x11},
+                Reloc::Debug { size: 4, addend: 0x19},
+            )?;
+            // comp dir
+            obj.link_with(
+                Link { from: ".debug_info", to: ".debug_str", at: 0x15},
+                Reloc::Debug { size: 4, addend: 0x7},
+            )?;
+            // var name
+            obj.link_with(
+                Link { from: ".debug_info", to: ".debug_str", at: 0x1a},
```

Comment:
It might be worth considering "enumifying" this for the debug_info, debug_str, etc., instead of strings (which have higher chance of typos); for now probably not worth it?

Reply:
I wouldn't want to limit the allowable debug sections by using an enum. What I think does make sense is for all declarations (function/variable/debug-section) to return a declaration id, and future references to those declarations (such as in definitions and links) should be via the id, not the name. cranelift-module already does this, but it could be extended down into faerie.